llvm/lib/Transforms/FuzzIntrospector/FuzzIntrospector.cpp

```diff
@@ -1132,3 +1184,234 @@ static RegisterStandardPasses
+  size_t trueComp = 0, falseComp = 0;
+
+  // iterate and skip those reachable by false side
+  for (auto BB: trueReachable) {
+    if (falseReachable.find(BB) != falseReachable.end()) {
+      continue;
+    }
+
+    if (BBComplexityMap.find(BB) == BBComplexityMap.end()) {
+      BBComplexityMap[BB] = calculateBBComplexity(BB);
+    }
+    trueComp += BBComplexityMap[BB];
+  }
+
+  // iterate and skip those reachable by true side
+  for (auto BB: falseReachable) {
+    if (trueReachable.find(BB) != trueReachable.end()) {
+      continue;
+    }
+
+    if (BBComplexityMap.find(BB) == BBComplexityMap.end()) {
+      BBComplexityMap[BB] = calculateBBComplexity(BB);
+    }
+    falseComp += BBComplexityMap[BB];
+  }
+
+  return make_pair(trueComp, falseComp);
+}
+
+std::pair<std::string, std::string> FuzzIntrospector::getInsnDebugInfo(Instruction *I) {
+  std::string ret_string = "", ret_line = "";
+
+  DILocation *Loc = I->getDebugLoc();
+  if (Loc != NULL)
+  {
+    ret_line = std::to_string(Loc->getLine());
+    ret_string = Loc->getFilename().str() +
+              +":" + ret_line + "," + std::to_string(Loc->getColumn());
+  }
+  else {
+    logPrintf(L1, "No debug info!!\n");
+  }
+
+  return make_pair(ret_string, ret_line);
+}
+
+std::pair<std::string, std::string> FuzzIntrospector::getBBDebugInfo(BasicBlock *BB, DILocation *prevLoc) {
+  std::pair<std::string, std::string> result = make_pair("", "");
+
+  BasicBlock *currBB = BB;
+  BranchInst *currBI;
+  Instruction *currTI, *currI;
+  DILocation *currLoc;
+
+  do {
+    currTI = currBB->getTerminator();
+    currI = currBB->getFirstNonPHIOrDbgOrLifetime(true);
+    if (currI == nullptr)
+      break;
+    currLoc = currI->getDebugLoc();
+    currBI = dyn_cast<BranchInst>(currTI);
+    if (currBI && !currBI->isConditional()){
+      currBB = currBI->getSuccessor(0);
+    }
+    else
+      break;
+  } while (currLoc == prevLoc);
+
+  if (currI)
+    result = getInsnDebugInfo(currI);
+  // for (auto &I: *BB) {
+  //   result = getInsnDebugInfo(&I);
+  //   if (result.first.length() > 0) {
+  //     break;
+  //   }
+  // }
+  return result;
+}
+
+// void FuzzIntrospector::writeOutMap(std::map<std::string, std::vector<std::pair<std::string, size_t>>> outMap, std::string fileName) {
+// void FuzzIntrospector::writeOutMap(std::map<std::string, BranchSidesComplexity> outMap, std::string fileName) {
+void FuzzIntrospector::writeOutMap(std::vector<BranchProfileEntry> outMap, std::string fileName) {
+  std::error_code EC;
+  logPrintf(L1, "Logging branchProfile to %s\n", fileName.c_str());
+
+  auto YamlStream = std::make_unique<raw_fd_ostream>(
+      fileName, EC, llvm::sys::fs::OpenFlags::OF_None);
+  yaml::Output YamlOut(*YamlStream);
+
+  YamlOut << outMap;
+}
+
+// Return cyclomatic complexity of called functions in the BB
+size_t FuzzIntrospector::calculateBBComplexity(BasicBlock *BB) {
```

Comment:
In the context of this method, do we need it? I'm thinking from the perspective of accummulating the cyclomatic complexity of an arbitrary set of functions is really easy in the post-processing as we have all the data there. It might be nicer to keep the LLVM module a bit more lean and, instead of accumulating cyclomatic complexity only output the functions callable by a given basic block?

Perhaps even more than that we might extract all of the callees for each basic block and then output this? I have a feeling this sort of data might be useful for other post-processing tasks?

In general, could we decrease some of the logic in the compiler plugin and do more of the reasoning in the post-processing? Down the line this might also be nice for example for doing the same technique for other languages?

I think the same principle applies to `findComplexities` -- it does a bunch of computation and then uses that as data in the data collector, but could we somehow push the computation out of the compiler plugin and into the post-processing part, and then use the compiler plugin as more of a data extractor?

Reply:
Thanks for the comments. Actually once I have these implemented and tested locally I started to come up with different approaches to profile branches (please check the latest changes in the design doc I shared with you). In a nutshell I totally agree with delegating computation to post-process, and probably I will deprecated these logic from LLVM side and only extract the list of functions called by each branch side. This is the approach that I am more inclined with as providing more raw data gives such more flexibility on the types of analysis we can perform later in post-process.

Reply:
> please check the latest changes in the design doc I shared with you

Will get this done tomorrow morning -- apologies for the delay here.